clustermesh: add mcs-api CRDs auto-installation

MrFreezeex opened the pull request:
This PR add MCS-API CRDs auto installation just like Cilium CRDs based on the existing infra to register CRDs (kudos to those who made this very extensible!). This together with https://github.com/cilium/cilium/pull/40506 will allows for 0 additional step for MCS-API install on top of regular clustermesh install in the future (probably Kubernetes >= 1.36 for running the correct CoreDNS version) and also ease a bit running MCS-API conformance test in our CI (probably soon after this will be merged :eyes:!)

```release-note
clustermesh: add MCS-API CRDs auto-installation
```

MrFreezeex committed:
clustermesh: gateway-api: ignore MCS-API CRD check if auto-installing

If we are installing MCS-API CRDs gateway-api and clustermesh MCS-API
integration should not check for the CRDs presence. This is similar to
how other CRDs are done except that we should keep checking in case this
auto installation is disabled (or even if Cilium MCS-API support is not
even enabled in the Gateway-API case).

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

MrFreezeex committed:
docs: document MCS-API CRD auto-install

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

MrFreezeex committed:
clustermesh: add config for auto installing MCS-API CRDs

Add clustermesh-mcs-api-crds config defaulting to true that will
be auto installing MCS-API CRDs. Also wire it up in helm configuration.

This commit only adds the config, the actual code to install CRDs will
be added in a future commit.

Also rename the existing field to enable MCS-API support to align on the
new short name to install CRDs.

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

MrFreezeex committed:
clustermesh: move MCS-API config to types package

Move MCSAPIConfig to types package to prevent future dependency cycle.
Also introduce DefaultMCSAPIConfig symbol at the same time.

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

giorio94 commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -3335,6 +3335,8 @@ clustermesh:
     # -- Enable Multi-Cluster Services API support
     enabled: false
+    # -- Enabled MCS-API CRDs auto-installation
+    installCRDs: true
```
I'm personally a bit on the fence on whether this should be true by default. Maybe better to have it false for the moment, similarly to `corednsAutoConfigure`, just to be on the safe side? I don't have a strong opinion though.

MrFreezeex replied:
I would prefer to keep it enabled by default because while `corednsAutoConfigure` may not work especially in cloud providers situation I don't see when that would be not working :thinking: even if some users might want to install CRDs before cilium-operator is started I don't think it's a bad thing that it enforce the CRD version that we support and even with that users can just opt-out anyway

giorio94 replied:
My main concern is that it may downgrade the version manually installed by users (as that would not have the expected labels), potentially completely breaking existing resources.

MrFreezeex replied:
Well right now we never told users to upgrade the CRDs and apart from this: https://github.com/kubernetes-sigs/mcs-api/commit/c635d41df2f3eefab19646722b70995bc122c5b9 which is a very minor change we are at the latest version so there's not really a downgrade possible right now at least.

It's possible that in the future things like that could happen but users would not gain any actual features to upgrade to something not supported since usually all the integration with MCS-API go through the ServiceImport and we fully control those fields and for ServiceExport we won't be reading fields that we do not support yet.

giorio94 replied:
Yeah, I know they would not benefit from an upgrade, but they may still install a different version for whatever reason (I'm mostly referring to when there will be the v1beta1 version), and it would be bad if they lost all their ServiceExport resources because of an unexpected downgrade to v1alpha1. Not saying it will happen, just to make sure we don't cause unexpected issues in these cases. 

That said, I totally agree that it is way more convenient for the vast majority of use-cases if Cilium just manages the CRDs by itself, so I'm fine with leaving this enabled if we are confident enough it won't cause surprises.

giorio94 commented on pkg/k8s/synced/crd.go:
```diff
@@ -120,13 +126,23 @@ func GatewayAPIResourceNames() []string {
+		res = append(res, MCSAPIResourceNames()...)
+	}
 	return res
 }
```
I think these changes could be reverted. If I'm not missing anything, the only place this would apply is in the test helpers, but that's for the legacy Ginkgo tests, so we can just ignore them unless you encountered issues (I would not expect these CRDs to be ever installed in those test envs).

giorio94 commented on pkg/clustermesh/mcsapi/crdinstall.go:
```diff
@@ -0,0 +1,126 @@
+)
+
+// RegisterCRDs registers MCS-API CRDs with the K8s apiserver.
+func RegisterCRDs(logger *slog.Logger, clientset client.Clientset) error {
```
Nit. A few methods are exported, while others are unexported, even though they are all only used internally. It could make sense to uniform them.

giorio94 commented on pkg/clustermesh/mcsapi/crdinstall.go:
```diff
@@ -0,0 +1,126 @@
+	name string,
+	crd apiextensionsv1.CustomResourceDefinition,
+) *apiextensionsv1.CustomResourceDefinition {
+	res := crd.DeepCopy()
```
Nit. `DeepCopy` is not necessary, as the original object seems never used.

giorio94 commented on pkg/clustermesh/mcsapi/crdinstall.go:
```diff
@@ -0,0 +1,126 @@
+	res := crd.DeepCopy()
+	res.Labels = map[string]string{
+		k8sconst.CustomResourceDefinitionSchemaVersionKey: k8sconst.CustomResourceDefinitionSchemaVersion,
+	}
```
Is `Labels` guaranteed to never be nil?

MrFreezeex replied:
It isn't technically guaranteed but there is no label on the CRD so far at least: https://github.com/kubernetes-sigs/mcs-api/blob/master/config/crd/multicluster.x-k8s.io_serviceexports.yaml

giorio94 replied:
I guess one could be on the safe side by adding the new label (creating the map if nil), rather than overriding all labels.

giorio94 commented on pkg/clustermesh/mcsapi/crdinstall.go:
```diff
@@ -0,0 +1,126 @@
+			constructV1CRD(crdMetaName, crd),
+			crdhelpers.NewDefaultPoller(),
+			k8sconst.CustomResourceDefinitionSchemaVersionKey,
+			versioncheck.MustVersion(k8sconst.CustomResourceDefinitionSchemaVersion),
```
I think it would be better to use a different `version` key, as there would be no guarantee that this one gets incremented if one of the MCS-API CRDs are updated (the increment is manual, while the update would be automatic on dependency update), which would lead to broken (i.e., missed) updates. Does the MCS-API package provide some version information that could be used?

MrFreezeex replied:
My reasoning is that we should very rarely backport a CRD update I think and IIUC this version is incremented by default on every release so it should already work without us doing anything specific here. Although that's true that when we do a backport here we could miss something, WDYT?

giorio94 replied:
> IIUC this version is incremented by default on every release so it should already work without us doing anything specific here

I'm not that sure if that't actually true at the moment. However, there may also be some changes in this area to skip updating CRDs when not actually necessary, as that can have a significant impact on the Kubernetes API Server due to the restart of all watchers: https://cilium.slack.com/archives/C069SNDMLA2/p1750869851447259

That said, even if done at release time, it would be still a problem (and lead to confusing behavior) if you don't upgrade to a released version, such as during local testing, or in case of upgrade/downgrade CI tests.

MrFreezeex replied:
Ah yep right I didn't gave much though about CI/testing it would be better then indeed! I added a new symbol version but I am not entirely sure how we would keep it up to date with renovate in mind :thinking:

giorio94 replied:
Ideally that should be something in the upstream repo, otherwise it would not make much difference whether it is the same or a different symbol, the bump would still need to be performed manually, and it would be easy to miss.

giorio94 replied:
Otherwise I guess that one possibility could be computing say an hash of the two CRDs, which would work great to determine if they are equal or different, but it would not give ordering to prevent downgrades. Maybe one middle ground could be using the actual CRD version + the hash, to avoid downgrading once the version is bumped, but that may require some changes to the CRDs installation library to ensure that it works.

MrFreezeex replied:
I will try to add that in the mcs-api repo let's see :eyes:

Putting this back to draft meanwhile!

giorio94 commented on Documentation/network/clustermesh/mcsapi.rst:
```diff
@@ -54,6 +46,10 @@ configure CoreDNS manually, you need to execute the following steps:
 
    .. code-block:: shell-session
 
+      # Install MCS-API CRDs if Cilium is not yet started or CRDs automatic CRDs installation is disabled (optional)
```
```suggestion
      # Install MCS-API CRDs if Cilium has not yet started or automatic CRDs installation is disabled (optional)
```

What does `optional` mean in this context?

MrFreezeex replied:
It refers to the if earlier in the sentence meaning that if Cilium is already installing the CRD they would not need this step

giorio94 replied:
Ah, to me that is already implied by the `if ...`, that's why I was a bit confused.

MrFreezeex replied:
Maybe I can put that before this sentence that might be more clear?

giorio94 reviewed (commented):
Thanks, overall makes sense to me. Just a couple of comments/questions inline.

It would be good to have some way (likely with renovate) to automatically keep the reference in the docs aligned with the embedded one. But it can definitely be a follow-up item IMO.

Artyop commented on test/helpers/kubectl.go:
```diff
@@ -62,8 +63,8 @@ const (
 
-	// CIIntegrationAKS contains the constants to be used when running tests on AKS.
-	CIIntegrationAKS = "aks"
+	// CIIntegrationAKS contains the constants to be used when running tests on ASK.
```
Little typo here? as well as in the comment

MrFreezeex replied:
Woops yep that's my codespell in my editor and I forgot to not include those :x

Artyop commented on test/helpers/kubectl.go:
```diff
@@ -32,6 +32,7 @@ import (
 	v1 "k8s.io/api/core/v1"
 
 	"github.com/cilium/cilium/api/v1/models"
+	mcsapitypes "github.com/cilium/cilium/pkg/clustermesh/mcsapi/types"
```
This import looks unused

Artyop reviewed (commented):
few nits

MrFreezeex committed:
k8s: allow to customise CRD version checking during CRD install

Wrap existing version check in a function call so that callers would be
able to use those helpers with a different need upgrade logic.

Also improve a bit the code to clarify the CRD variable name (currentCRD
and targetCRD instead of crd and targetCRD) and add crdName in
logfields.

Signed-off-by: Arthur Outhenin-Chalandre <git@mrfreezeex.fr>

MrFreezeex committed:
clustermesh: add MCS-API auto CRDs install

Add MCS-API auto CRD install based on the existing CRDs registration infra.

Signed-off-by: Arthur Outhenin-Chalandre <git@mrfreezeex.fr>

HadrienPatte reviewed (approved):
vendor LGTM

qmonnet commented on Documentation/operations/upgrade.rst:
```diff
@@ -316,7 +316,8 @@ communicating via the proxy must reconnect to re-establish connections.
   This may change in future. See :gh-issue:`35823`
   and :gh-issue:`17177` for further discussion on this topic.
-* MCS-API CRDs need to be updated, see the MCS-API :ref:`clustermesh_mcsapi_prereqs` for updated CRD links.
+* If MCS-API support is enabled, Cilium now install and manage MCS-API CRDs by default.
```
```suggestion
* If MCS-API support is enabled, Cilium now installs and manages MCS-API CRDs by default.
```

MrFreezeex replied:
Woops indeed, thanks!

qmonnet reviewed (changes requested):
Please fix the grammar nit for the upgrade note. Doc changes look good otherwise, thanks!

qmonnet reviewed (approved):
Doc update looks good, thank you!

giorio94 commented on pkg/clustermesh/mcsapi/types/config.go:
```diff
@@ -0,0 +1,27 @@
+	flags.Bool(
+		"clustermesh-enable-mcs-api",
+		cfg.ClusterMeshEnableMCSAPI,
+		"Whether or not the MCS API support is enabled.",
```
Nit. "Enable Cluster Mesh MCS API support" would be a bit more direct and better aligned with the wording of the other flags (I know that you just copy/pasted it, but figured I'd still leave a comment).

giorio94 commented on pkg/clustermesh/mcsapi/types/config.go:
```diff
@@ -24,4 +28,14 @@ func (cfg MCSAPIConfig) Flags(flags *pflag.FlagSet) {
+	flags.Bool(
+		"clustermesh-install-mcs-api-crds",
+		cfg.ClusterMeshInstallMCSAPICRDs,
+		"Whether or not the MCS API CRDs should be automatically installed. "+
```
Nit. Prefer a more direct phrasing. E.g., "Install and manage the MCS API CRDs. Only applicable if MCS API support is enabled."

giorio94 commented on pkg/clustermesh/mcsapi/types/config.go:
```diff
@@ -24,4 +28,14 @@ func (cfg MCSAPIConfig) Flags(flags *pflag.FlagSet) {
 		"Whether or not the MCS API support is enabled.",
 	)
+	flags.Bool(
+		"clustermesh-install-mcs-api-crds",
```
Nit. `clustermesh-mcs-api-install-crds` would group better with other related options possibly introduced in the future.

MrFreezeex replied:
Ah yep good point! I also renamed the existing MCS-API option so that it's grouped with this one :D

giorio94 commented on pkg/k8s/apis/cilium.io/client/register.go:
Nit Typo in the commit title (`s/durind/during`).

giorio94 commented on pkg/clustermesh/mcsapi/crdinstall.go:
```diff
@@ -0,0 +1,129 @@
+
+// createCRD creates and updates a CRD.
+// It should be called on operator startup but is idempotent and safe to call again.
+func createCRD(logger *slog.Logger, crdVersionedName string) func(clientset apiextensionsclient.Interface) error {
```
Nit. I'd simplify this to avoid returning a function, given that both are invoked at the same time. 

It could also make sense to have a dedicated struct to group all these functions, and avoid having to pass common parameters around. But that's mostly personal preference, feel free to leave as is.

MrFreezeex replied:
Makes sense! I was mostly copying the existing code which is similar, but I will apply your comments to the existing code then!

giorio94 commented on pkg/clustermesh/mcsapi/crdinstall.go:
```diff
@@ -0,0 +1,129 @@
+)
+
+// registerCRDs registers MCS-API CRDs with the K8s apiserver.
+func registerCRDs(logger *slog.Logger, clientset client.Clientset) error {
```
Nit. Could be merged into the function below.

giorio94 commented on pkg/clustermesh/mcsapi/crdinstall.go:
```diff
@@ -0,0 +1,129 @@
+	}
+}
+
+func NeedsUpdateMCS(crd, clusterCRD *apiextensionsv1.CustomResourceDefinition) bool {
```
Nit. Any reason why this is exported, while the others are not?

MrFreezeex replied:
No reason I probably did it to mimic the NeedsUpdate func used for Cilium CRD without thinking too much about it!

giorio94 commented on pkg/clustermesh/mcsapi/crdinstall.go:
```diff
@@ -0,0 +1,129 @@
+		crdBytes []byte
+	)
+
+	logAttr := slog.String("crdName", crdName)
```
Log KVs should be inlined according to the [linter](https://github.com/cilium/cilium/blob/main/.golangci.yaml#L122-L124), and key names should be constants (typically in the logfields package). For some reason it is not complaining here, but I'd suggest aligning it anyways.

giorio94 commented on pkg/clustermesh/mcsapi/crdinstall.go:
```diff
@@ -0,0 +1,129 @@
+	case mcsapiv1alpha1.ServiceExportVersionedName:
+		crdBytes = mcsapicrd.ServiceExportCRD
+	default:
+		logging.Fatal(logger, "Pregenerated CRD does not exist", logAttr)
```
Could avoid panicking, given that the caller can return an error.

MrFreezeex replied:
Since the cilium CRD code does the same and that it isn't likely to go in those error conditions as the CRDs list is a few lines above and quite short I didn't do this but I would be happy to do it if you insist (and also probably doing it in the Cilium CRD code then)!

giorio94 commented on pkg/clustermesh/mcsapi/crdinstall.go:
```diff
@@ -0,0 +1,129 @@
+	clusterRevision, err := strconv.Atoi(rev)
+	revision, errAtoi := strconv.Atoi(crd.Labels[mcsapicrd.CustomResourceDefinitionSchemaRevisionLabel])
+	if errAtoi != nil {
+		panic(fmt.Sprintf("invalid CRD revision label on CRD %s: %s", crd.Name, crd.Labels[mcsapicrd.CustomResourceDefinitionSchemaRevisionLabel]))
```
The panic here may raise some concerns (and similarly the `MustVersion` above) even though I agree that you are acting on the embedded CRD object, which is known a-priory. May be worth replacing it with a log error, just to be on the safe side (or let `NeedUpdateCRDFunc` return an error).

MrFreezeex replied:
Makes sense, I did that in case the CRD from upstream get updates in a way we don't like here

giorio94 commented on pkg/clustermesh/mcsapi/cell.go:
```diff
@@ -154,3 +157,21 @@ func registerMCSAPIController(params mcsAPIParams) error {
+	cell.Out
+
+	Func apis.RegisterCRDsFunc `group:"register-crd-funcs"`
+}
```
Nit. Could export the one in `pkg/k8s/apis/cell.go`, instead of creating a new one.

giorio94 commented on pkg/k8s/apis/crdhelpers/register.go:
```diff
@@ -22,6 +22,8 @@ import (
 	"github.com/cilium/cilium/pkg/versioncheck"
 )
 
+type NeedUpdateCRDFunc func(crd, clusterCRD *apiextensionsv1.CustomResourceDefinition) bool
```
Nit. Maybe `target` and `current` (or something like these) would be a bit more expressive as parameter names.

MrFreezeex replied:
I agree, I updated the cilium CRD code which is similar to this too

giorio94 reviewed (commented):
Thanks! Overall looks good to me, a few minor nits inline.

giorio94 commented on pkg/clustermesh/mcsapi/types/config.go:
```diff
@@ -8,20 +8,33 @@ import (
 func (cfg MCSAPIConfig) Flags(flags *pflag.FlagSet) {
 	flags.Bool(
-		"clustermesh-enable-mcs-api",
+		"clustermesh-mcs-api-enable",
```
This is technically a breaking change, which would need to be handled according to the usual policy. I'd suggest keeping it as is for the moment, to not extend the scope of this PR, and possibly change it in a separate one (I don't think it is strictly needed though, given that this is the one to enable the feature).

MrFreezeex replied:
Ah ok I see, then indeed let's keep it as is!

giorio94 commented on pkg/clustermesh/mcsapi/types/config.go:
```diff
@@ -8,20 +8,33 @@ import (
+		"clustermesh-mcs-api-enable",
 		cfg.ClusterMeshEnableMCSAPI,
-		"Whether or not the MCS API support is enabled.",
+		"Enable Cluster Mesh MCS-API support",
```
Nit. I'd probably move it to the previous commit.

giorio94 commented on pkg/clustermesh/mcsapi/types/config.go:
```diff
@@ -8,20 +8,33 @@ import (
+		"Enable Cluster Mesh MCS-API support",
 	)
+	flags.Bool(
+		"clustermesh-mcs-api-crds",
```
It should be `clustermesh-mcs-api-install-crds`

giorio94 commented on pkg/clustermesh/mcsapi/types/config.go:
```diff
@@ -8,20 +8,33 @@ import (
+	ClusterMeshEnableMCSAPI bool `mapstructure:"clustermesh-mcs-api-enable"`
+	// ClusterMeshInstallMCSAPICRDs control whether to automatically install the MCS API CRDs
+	// conditional on ClusterMeshEnableMCSAPI being true
+	ClusterMeshInstallMCSAPICRDs bool `mapstructure:"clustermesh-mcs-api-crds"`
```
Nit. The name of the field could be changed to `InstallCRDs` (or something like that), to reduce redundancy.

MrFreezeex replied:
I also removed the `ClusterMesh` prefix from the enable option for consistency then, is that ok or you prefer to keep this option as is?

giorio94 replied:
Sure, works for me. For the future (no need to change it now), I'd suggest doing a change like that in a separate commit, given that orthogonal to the other changes.

giorio94 reviewed (changes requested):
Thanks! A couple more comments inline.

sayboras reviewed (approved):
Changes look good for files owned by servicemesh team

tommyp1ckles commented on pkg/clustermesh/mcsapi/crdinstall.go:
```diff
@@ -0,0 +1,120 @@
+}
+
+func needsUpdateMCS(targetCRD, currentCRD *apiextensionsv1.CustomResourceDefinition) (bool, error) {
+	if currentCRD.Spec.Versions[0].Schema == nil {
```
should we guard against empty versions spec?

MrFreezeex replied:
Apparently the apiserver refuses such type of object, from a dumb test against a kube cluster I get this type of error:
`* spec.versions: Invalid value: []apiextensions.CustomResourceDefinitionVersion(nil): must have exactly one version marked as storage version`.

This specific check was copied from the existing code that update Cilium CRD so I kind of took it for granted, but looks fine after all :sweat_smile:.
